arch/armv7m/tls.S

```diff
@@ -0,0 +1,30 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * tls.S
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Lukasz Leczkowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#define __ASSEMBLY__
+
+.thumb
+.syntax unified
+
+
+.globl __aeabi_read_tp
+.type __aeabi_read_tp,%function
+/* __aeabi_read_tp must not clobber any registers except r0 */
+__aeabi_read_tp:
+	push {r1-r3, lr}
+	blx aeabi_read_tp_internal
+	pop {r1-r3, lr}
+	bx lr
+.size __aeabi_read_tp, .-__aeabi_read_tp
```

Comment:
Why not whole function in assemler?

Reply:
Because we need access to a variable in .data - on armv7m (and NOMMU in general) we use PIC/PIE, so all .bss and .data accesses are doubly indirect - via .got table. We wouldn't be able to do this in asm - we don't know .got table index for this variable.

Other feaseable option is to pass ready pointer to asm, but we need wrapper outside to save r1-r3 from being overwritten (compiler chose r3 for .got access operation, as we determined).

Reply:
**A bit long**

A fragment of the file `tls_internal.S` generated by the compiler:
```
aeabi_read_tp_internal:
.LFB0:
	.file 1 "tls_internal-1.c"
	.loc 1 31 1 view -0
	.cfi_startproc
	@ args = 0, pretend = 0, frame = 0
	@ frame_needed = 0, uses_anonymous_args = 0
	@ link register save eliminated.
	.loc 1 32 2 view .LVU1
	ldr	r3, .L3
	ldr	r3, [r9, r3]
	ldr	r3, [r3]
	.syntax unified
@ 32 "tls_internal-1.c" 1
	mov r0,r3
@ 0 "" 2
	.loc 1 36 1 is_stmt 0 view .LVU2
	.thumb
	.syntax unified
	bx	lr
.L4:
	.align	2
.L3:
	.word	.LANCHOR0(GOT)
	.cfi_endproc
.LFE0:
	.size	aeabi_read_tp_internal, .-aeabi_read_tp_internal
	.global	arm_tls_ptr
	.section	armtls,"aw"
	.align	2
	.set	.LANCHOR0,. + 0
	.type	arm_tls_ptr, %object
	.size	arm_tls_ptr, 4
arm_tls_ptr:
	.space	4
	.text
```
Besides, the function is defined as 
```
void *aeabi_read_tp_internal(void)
```
and there is no return statement, compiler warning:
```
warning: control reaches end of non-void function [-Wreturn-type]
```

Then maybe such solution in the assembler:
```
.thumb
.syntax unified

.align 2
.thumb_func
.globl __aeabi_read_tp
.type __aeabi_read_tp,%function
__aeabi_read_tp:
	ldr r0, 1f
	ldr r0, [r9, r0]
	ldr r0, [r0]
	bx lr
	.align 2
1:
	.word arm_tls_ptr(GOT)
.size __aeabi_read_tp, .-__aeabi_read_tp

.section armtls, "aw"
.align 2
.globl arm_tls_ptr
.type	arm_tls_ptr, %object
arm_tls_ptr:
	.word 0x0
/*
instead
	.word 0
can be
	.space 4
*/
.size	arm_tls_ptr, 4
```
and disassembly:
```
00000000 <__aeabi_read_tp>:
   0:   4802            ldr     r0, [pc, #8]    ; (c <__aeabi_read_tp+0xc>)
   2:   f859 0000       ldr.w   r0, [r9, r0]
   6:   6800            ldr     r0, [r0, #0]
   8:   4770            bx      lr
   a:   bf00            nop
   c:   00000000        .word   0x00000000      c: R_ARM_GOT32  arm_tls_ptr
```

Reply:
I like that compiler does this PIC/PIE stuff "behind the scenes". Right now this code will work regardless if it's PIC/PIE or if GOT pointer is in R9 or R8, compiler handles this stuff without us hardcoding anything. Because of that I opt to leave this as is.

Reply:
In that case, why not simpler:
```
void *aeabi_read_tp_internal(void)
{
	return (void *)arm_tls_ptr;
}
```

Reply:
Huh, yeah, haven't thought about it, good idea

Reply:
What do you think about this solution:
```
void *read_tls_ptr(void)
{
	return (void*)arm_tls_ptr;
}

void __attribute__((naked)) *__aeabi_read_tp_t5(void)
{
	__asm__ volatile(
		"push {r1-r4,r12,lr};"
		"bl read_tls_ptr;"
		"pop {r1-r4,r12,pc}");
}
```